keps/sig-multicluster/1645-multi-cluster-services-api/README.md

```diff
@@ -0,0 +1,925 @@
+  They can't be merged with other exports, and it seems like it would only
+  complicate deployments by even attempting to stretch them across clusters.
+  Instead, regular `ExternalName` type `Services` should be created in each
+  cluster individually.
+
+### Consumption of EndpointSlice
+
+To consume a supercluster service, users will use the domain name associated
+with their `ServiceExport`. When the mcsd-controller sees a `ServiceExport`, a
+`ServiceImport` will be introduced, which can be largely ignored by the user. 
+
+An `ServiceImport` is a service that may have endpoints in other clusters.
+This includes 3 scenarios:
+1. This service is running entirely in different cluster(s)
+1. This service has endpoints in other cluster(s) and in this cluster
+1. This service is running entirely in this cluster, but is exported to other cluster(s) as well
+
+For each exported service, one `ServiceExport` will exist in each cluster that
+runs the service. The mcsd-controller will create and maintain a derived
+`ServiceImport` in each cluster within the supercluster (see: [constraints and
+conflict resolution](#constraints-and-conflict-resolution)). If all `ServiceExport` instances are deleted, each
+`ServiceImport` will also be deleted from all clusters.
+
+Since a given `ServiceImport` may be backed by multiple `EndpointSlices`, a
+given `EndpointSlice` will reference its `ServiceImport` using the label
+`multicluster.kubernetes.io/imported-service-name` similarly to how an
+`EndpointSlice` is associated with its `Service` in a single cluster. Each
+imported `EndpointSlice` will also have a
+`multicluster.kubernetes.io/source-cluster` label with a registry-scoped unique
+identifier for the cluster.
+
+```golang
+// ServiceImport declares that the specified service should be exported to other clusters.
+type ServiceImport struct {
+ metav1.TypeMeta `json:",inline"`
+ // +optional
+ metav1.ObjectMeta `json:"metadata,omitempty"`
+ // +optional
+ Spec ServiceImportSpec `json:"spec,omitempty"`
+}
+
+// ServiceImportSpec contains the current status of an imported service and the
+// information necessary to consume it
+type ServiceImportSpec struct {
+  // +patchStrategy=merge
+  // +patchMergeKey=port
+  // +listType=map
+  // +listMapKey=port
+  // +listMapKey=protocol
+  Ports []ServicePort `json:"ports"`
+  // +optional
+  // +patchStrategy=merge
+  // +patchMergeKey=cluster
+  // +listType=map
+  // +listMapKey=cluster
+ Clusters []ClusterSpec `json:"clusters"`
+ // +optional
+ IPFamily corev1.IPFamily `json:"ipFamily"`
+ // +optional
+ IP string `json:"ip,omitempty"`
+}
+
+// ClusterSpec contains service configuration mapped to a specific cluster
+type ClusterSpec struct {
+ Cluster string `json:"cluster"`
+ // +optional
+ // +listType=set
+ TopologyKeys []string `json:"topologyKeys"`
+ // +optional
+ PublishNotReadyAddresses bool `json:"publishNotReadyAddresses"`
+ // +optional
+ SessionAffinity corev1.ServiceAffinity `json:"sessionAffinity"`
```

Comment:
Could this be kind of strange for the user to have different session affinity on a per-cluster level?
How does the user reason about this behavior?
Is this a useful steady state to be in?

Reply:
This would definitely be strange and not a useful steady state, but one of the goals of this API is to allow incremental rollout of changes. Maintaining affinity of the exporting cluster makes it easy to reason about what is happening - however undesirable it may be - during the transition.

Alternatively, we could e.g. ignore session affinity unless all clusters agree, or have some more advanced "pick the best config" logic but then it becomes much harder to reason about behavior during the transition and users will need to understand our specific merging logic at all times

Reply:
FTR: Jeremy and team and I went back and forth on this forever, and this seemed least-bad to us.